### PR TITLE
[receiver/dockerstatsreceiver] Initial PR to onboard dockerstats onto mdatagen

### PR DESCRIPTION
**Pull request by jamesmoessis**
First of all, apologies for the size of the PR. Thankfully, most of it is generated code/docs or mock/test data. Additionally, there is no change to the behaviour, since I haven't enabled the new implementation. 

**Description:** 
I wanted the initial PR to be backwards compatible with the existing implementation. So, I've re-implemented the `scrape` function as `scrapeV2` (leaving the original) and testing them both alongside each other to ensure they produce the same metrics, and to give the reviewer confidence that feature parity is maintained. The bulk of the new implementation is all in `receiver_v2.go`.

I've defined the metrics to be recorded in `metadata.yaml`. With this, the new implementation (which isn't enabled yet) you can turn off and on each metric via the config, and all metrics have generated docs in `documentation.md`. 

I've also fixed two bugs: 
- When collecting data from multiple containers, it seems that the resource metrics aren't appended but overridden, so the last container to be collected is the only one that you see metrics for. I made a minor change to fix this. 
- Specifying an HTTP endpoint (not unix socket) for the docker engine resulted in an incorrect dialer being used. It means it was ignoring the endpoint I specified and actually querying my local docker engine's unix socket. I had to make a minor change the to internal docker client to pass it an option, to make sure the correct dialer is being used. 

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9794

**Testing:** 
`TestScrapes` tests both implementations of `scrape` and ensures they emit the same metrics for the same datasource. I have mocked the docker engine using the HTTP test server. 

**Documentation:** 
`mdatagen` generates documentation in `documentation.md`. 

**Discussion points:**
How should transition to using the new implementation by default? There are a few breaking changes that are needed to move the component forward here. With the two implementations, we could temporarily expose a config which allows users to stay on the "deprecated" implementation, giving them time to migrate, as we make the breaking changes to the new implementation of `scrape` only. Eventually, we delete the old implementation of `scrape` and a whole bunch of code with it. Would love thoughts and feedback here :)

### CODE REVIEW

**Comment by jamesmoessis** [comment]
Hello @rmfitzpatrick @MovieStoreGuy can you please review :)

File: internal/docker/docker.go
Comment (jamesmoessis) [inline internal/docker/docker.go:59]: I had to make this change to the internal docker client because like I said in the PR description, it was using the incorrect dialer.

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [inline receiver/dockerstatsreceiver/receiver.go:123]: This is the bug I found regarding overwriting resource metrics. Only the last container stats was kept.

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [inline receiver/dockerstatsreceiver/receiver.go:275]: One of the breaking changes I am suggesting to make in a later PR would be to change `read, write, etc.` into an attribute, so we wouldn't need so many entries here. Also way less generated code+docs.

File: receiver/dockerstatsreceiver/receiver.go
Comment (MovieStoreGuy) [inline receiver/dockerstatsreceiver/receiver.go:128]: This is more a stylistic choice but I would create a seperate file for the `resultv2` and associated changes since if the intention is to deprecate and remove that old functionality, it should be a clean and easy removal from git instead of line changes and import statements.

File: receiver/dockerstatsreceiver/receiver.go
Comment (MovieStoreGuy) [reply receiver/dockerstatsreceiver/receiver.go:275]: Is there any reason why these can't happen now?

File: receiver/dockerstatsreceiver/receiver.go
Comment (MovieStoreGuy) [inline receiver/dockerstatsreceiver/receiver.go:342]: In the event that `entry.Op` is not present in the map, there should be a way to surface that information. 

An error may be the best means of doing so or providing some debug logs that can be configured on collector start up. Having things silently fail is worrisome.

**Review by MovieStoreGuy — COMMENTED** [review]
There is just a few things I have comments on atm, I will come back to review this in further detail.

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/metadata.yaml:88]: It essentially means no unit. Like for a percentage you would use this unit. It's wrong, and this should be `{periods}`, but that change (and a bunch of others) is for another PR. Right now I am just recreating what is already there with mdatagen. I didn't want to mix up the conversations of "what are the correct metrics" and "how to re-implement", otherwise this PR would be way to huge and complicated.

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/metadata.yaml:3]: The goal of this PR is to match the functionality of the existing implementation, which is on `1.6.1`. I definitely want to use the latest semconv, but that is something I want to save for a later PR.

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/metadata.yaml:28]: It's the device reference in these kernel docs: https://www.kernel.org/doc/Documentation/cgroup-v1/blkio-controller.txt

Would "device major number for block IO operation" be specific enough?

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/metadata.yaml:41]: Yes it does. Looking at the `moby` code, this is what it says: 

```
// getSystemCPUUsage returns the host system's cpu usage in
// nanoseconds. An error is returned if the format of the underlying
// file does not match.
```

source: https://github.com/moby/moby/blob/master/daemon/stats/collector.go#L123

This also raises another change we should make, because this metric is reported per-container, needlessly creating timeseries. In a later PR I can fix that.

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/metadata.yaml:42]: Yes, I double checked it (see reply to other comment)

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/metadata.yaml:96]: "1" is used a lot in the existing implementation. Like I said, the goal of this PR is to recreate the existing implementation, then the later PRs will fix up the units and a few other things. 

I've made a list of the changes I want to make in the parent issue of this PR: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9794#issuecomment-1182616779

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go:64]: No it hasn't

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go:128]: Yeah, I like that idea. I will do that.

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go:275]: I don't want to change the existing metrics in this PR. I want to prove this implementation is equivalent output to the existing implementation, and then we can work on it from there. 

I've listed the improvements in this comment in the parent issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9794#issuecomment-1182616779

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go:195]: You're right, I removed them and the test passed. Thanks for pointing that out 👍

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go:342]: Yeah, good point. I think a debug log is appropriate. I will add one.

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go]: Done

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go]: Done

File: receiver/dockerstatsreceiver/receiver.go
Comment (MovieStoreGuy) [reply receiver/dockerstatsreceiver/receiver.go:61]: Claim your bug!

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go]: I've added the `const` you suggested. 

In terms of the resource detector, that would need a further look. The package you linked is the Go SDK, is there something in the collector that does this? Perhaps can look at it in another PR.

File: receiver/dockerstatsreceiver/receiver.go
Comment (rmfitzpatrick) [inline receiver/dockerstatsreceiver/receiver.go:70]: Would this affect the docker observer as well such that it'd be preferable to be in the internal `NewDockerClient()`?

File: receiver/dockerstatsreceiver/receiver_test.go
Comment (rmfitzpatrick) [inline receiver/dockerstatsreceiver/receiver_test.go:159]: Can this be broken out and only invoked within the v2 scrape methods instead of all tests?

File: receiver/dockerstatsreceiver/testdata/mock/single_container/container.json
Comment (rmfitzpatrick) [inline receiver/dockerstatsreceiver/testdata/mock/single_container/container.json:2]: nit: can the json files be sorted for potentially easier maintenance/review over time?

File: receiver/dockerstatsreceiver/receiver.go
Comment (rmfitzpatrick) [inline receiver/dockerstatsreceiver/receiver.go:51]: Can we propagate `ProvidePerCoreCPUMetrics` to the underlying per core cpu metrics enabled flag?

Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go:54]: Yeah, good idea. I've made it so if the user has that flag set to `true`, I override it on the mdatagen metric settings (which are false by default).
```go
if config.ProvidePerCoreCPUMetrics {
  config.MetricsConfig.ContainerCPUUsagePercpu.Enabled = config.ProvidePerCoreCPUMetrics
}
``` 

I've also enabled the per-cpu metrics in the test, so this behaviour is verified.

File: receiver/dockerstatsreceiver/testdata/mock/single_container/container.json
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/testdata/mock/single_container/container.json]: Done!

File: receiver/dockerstatsreceiver/receiver_test.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver_test.go]: Done!

File: receiver/dockerstatsreceiver/receiver.go
Comment (MovieStoreGuy) [reply receiver/dockerstatsreceiver/receiver.go:73]: I would need to double check but it is fair to say it would.

Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go:73]: Yeah it's a good question. I think it would, but I'm not 100% convinced that the solution here is necessarily the desired general solution for the client. 

I just tried pinning `go-connections` to `0.4.0` and that seems to fix it. I wonder if that is a better solution here.

**Comment by jamesmoessis** [comment]
Going to document this here and link to the comment from the exclusion in `go.mod`. 

There's a problem in v0.4.1 of go-connections/sockets.go where it overrides the proto and doesn't reset it when a new scheme is applied. This had the effect of making request to a unix socket even though an http endpoint was specified for the docker engine. 

There are two possible solutions: 
1. Pinning to 0.4.0 and excluding 0.4.1 
2. Passing in a default dialer to via opts

For this PR, I think option (1) is better.

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go]: Ok see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12322/commits/2709245f8aaa7b7d8906afe07e2dbab6c7a4f18a 

I've pinned to 0.4.0 and excluded 0.4.1. Potentially that should be done in the internal docker client package itself, but I think I will raise an issue for it and fix it separately from this PR. 

This is a more robust solution than checking if the endpoint starts with `http` and then overriding the dial context.

**Comment by mx-psi** [comment]
@jamesmoessis Note that, per [the ref doc](https://go.dev/ref/mod#go-mod-file-exclude)

> exclude directives only apply in the main module’s go.mod file and are ignored in other modules

So we need to add this:

1. on the main go.mod file
2. on the contrib releases manifest

**Comment by jamesmoessis** [comment]
Thanks @mx-psi for pointing that out. I have updated the main `go.mod`. I can also raise another PR in the releases repo. 

Before I do that, though, I would like to ask your opinion on whether this should be a `replace` or an `exclude`? Both seem to work: 

```
replace github.com/docker/go-connections v0.4.1-0.20210727194412-58542c764a11 => github.com/docker/go-connections v0.4.0
```
or
```
exclude github.com/docker/go-connections v0.4.1-0.20210727194412-58542c764a
```

**Comment by mx-psi** [comment]
@jamesmoessis I am not sure how `exclude` interacts with pseudoversions, so maybe a `replace` is indeed better.

File: receiver/dockerstatsreceiver/documentation.md
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/documentation.md:64]: Bold means it's emitted by default. This one is not emitted by default, as per the existing config defaults. There is further explanation at the bottom of the file.

File: internal/docker/docker.go
Comment (MovieStoreGuy) [inline internal/docker/docker.go:59]: If we have reverted the `go-connections` version, does this still need to happen?
Or is this just a safe nice to have?

File: receiver/dockerstatsreceiver/doc.go
Comment (MovieStoreGuy) [inline receiver/dockerstatsreceiver/doc.go:17]: Optional

nit: to keep this align with the existing files for `doc.go` declare the `//go:generate` line above the package .

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (MovieStoreGuy) [reply receiver/dockerstatsreceiver/metadata.yaml:28]: > It's the device reference in these kernel docs: https://www.kernel.org/doc/Documentation/cgroup-v1/blkio-controller.txt
> 
> Would "device major number for block IO operation" be specific enough?

Yeah, I think that is clearly for those who are reading for the first time.

**Review by MovieStoreGuy — APPROVED** [review]
I am happy with these changes, only nits that really shouldn't matter

File: receiver/dockerstatsreceiver/documentation.md
Comment (mx-psi) [reply receiver/dockerstatsreceiver/documentation.md:64]: Gotcha, thanks!

File: internal/docker/docker.go
Comment (jamesmoessis) [reply internal/docker/docker.go:59]: Yeah, it doesn't need to be there anymore, but imo it doesn't hurt to leave it in, and it could be nice to have for future purposes or similar issues.

**Comment by jamesmoessis** [comment]
@rmfitzpatrick @mx-psi I believe I have addressed all comments, is there anything else I can do to move this PR forward?

**Comment by mx-psi** [comment]
Will wait for @rmfitzpatrick to give a final review

**Review by rmfitzpatrick — APPROVED** [review]
Thanks!

**Review by mx-psi — APPROVED** [review]

**Comment by mx-psi** [comment]
@jamesmoessis Regarding

> How should transition to using the new implementation by default?

I would suggest using a [feature gate](https://github.com/open-telemetry/opentelemetry-collector/tree/main/service/featuregate#usage) for this

**Comment by jamesmoessis** [comment]
@mx-psi thankyou for the suggestion, I will implement in my next PR!

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go:47]: Yes, we should do that, and it's my eventual goal

**Pull request merged**
